src/durableorchestrationcontext.ts

```diff
@@ -30,7 +30,13 @@ import {
 } from "./task";
 import moment = require("moment");
 import { ReplaySchema } from "./replaySchema";

```

Comment:
The benefits of this change are not as clear-cut as I initially thought they were. Some feedback on the description:

1. I don't think the "Before" examples are accurate. I think most people (including our templates) would put `'sayHelloWithActivity'` in a `const` which is much safer.
1. It bothers me in the "After" example that `sayHelloWithActivity` is used before it's defined. I know technically it will work as-is, but some linters will flag that as a problem and from a readability perspective I agree with the linters. With a string the order doesn't matter, but with the object it does.
1. I would like us to explore more complex projects where these aren't all in the same file. My gut is telling me that the imports could get complicated now that we depend on the output of registering the handler instead of using a static string. For example - have you ever run into circular import errors? Not sure if they're possible in durable, but they're a huge pain haha

Reply:
Re: 1. Fair enough, I've updated the description.
Re: 2. We could switch the order in the templates, but I agree it is inconvenient.
Re: 3. I've tested this with the following project, and it worked fine:

```TS
// helloActivity.ts:
import * as df from 'durable-functions';
import { ActivityHandler } from 'durable-functions';

const helloActivityHandler: ActivityHandler = (input: string): string => {
    return `Hello, ${input}`;
};

export default df.app.activity('helloActivity', { handler: helloActivityHandler });

//

// sayHelloWithActivity.ts
import * as df from 'durable-functions';
import { OrchestrationContext, OrchestrationHandler } from 'durable-functions';
import helloActivity from './helloActivity';

const sayHelloWithActivityHandler: OrchestrationHandler = function* (context: OrchestrationContext) {
    const input = context.df.getInput();
    return yield context.df.callActivity(helloActivity, input);
};

export default df.app.orchestration('sayHelloWithActivity', sayHelloWithActivityHandler);

// 

// helloSequence.ts

import * as df from 'durable-functions';
import { OrchestrationContext, OrchestrationHandler } from 'durable-functions';
import sayHelloWithActivity from './sayHelloWithActivity';

const orchestrator: OrchestrationHandler = function* (context: OrchestrationContext) {
    const outputs = [];
    outputs.push(yield context.df.callSubOrchestrator(sayHelloWithActivity, 'Tokyo'));
    outputs.push(yield context.df.callSubOrchestrator(sayHelloWithActivity, 'Seattle'));
    outputs.push(yield context.df.callSubOrchestrator(sayHelloWithActivity, 'Cairo'));

    return outputs;
};
df.app.orchestration('helloSequence', orchestrator);
```

I'd argue I think it's cleaner to use this style when you have multiple orchestrations/activities across multiple files, and might be cleaner to use constant strings at the top if everything is in one file.

Reply:
I feel these patterns are pretty close in functionality, and largely come down to personal preference. In that case, is the downside of overloads really worth the new option? If the new one was clearly better, I'd be fine with the overloads because the downside would be temporary. As it is, I feel like we'll be stuck supporting both overloads indefinitely because there will never be a compelling enough reason to remove the old pattern.

To remind y'all, this is why I try to avoid overloads - it turns a one line error into an ugly five line error:

|Without overloads|With overloads|
|---|---|
|<img width="804" alt="Screenshot 2023-02-08 at 3 51 22 PM" src="https://user-images.githubusercontent.com/11282622/217696794-fb548315-0b6e-4642-a78a-9bf7a93dc8a1.png">|<img width="695" alt="Screenshot 2023-02-08 at 3 48 18 PM" src="https://user-images.githubusercontent.com/11282622/217696790-1ddd68e8-6310-4da2-b084-22d105162670.png">|

@davidmrdavid @hossam-nasr Are you both confident you like the new pattern better? If yes, I will defer to you guys and I'm fine getting it in. If no, I will remind you that we have plenty of time to tinker with this design since we're doing it in a non-breaking way. We can easily ship this feature any time in the future regardless of announcement dates

Reply:
@ejizba: Thanks for the concrete example. I'm confident that, across all DF SDKs, we're looking to support safer ways of calling Activities/Sub-Orchestrators/Entities by providing alternatives to this string-based parameter. I know I've been asked to do something equivalent to this in Python's new progmodel, although overloads there a bit a nicer in error cases.

Yes, we have time to tinker with this design and it appears we would be ok to merge this at a later date, so I'm no rush. I think this would be something I'd love to demo, among other bits, to the DF team. That would help us be in the same page about the benefits of the design.

Reply:
I don't have strong feelings about either merging this now (and including it in preview), or punting until later, after/during preview. I will defer to @davidmrdavid on this one. Let me know if you think we should have this PR in before preview and I'll happily rebase the branch so we can merge.